data table tutorial (#304)
reflex-dev/reflex-web: docs/datatable-tutorial/add_interactivity.md

```python exec
import reflex as rx
from pcweb.base_state import State
from pcweb.templates.docpage import docdemobox
from pcweb.pages.docs.datatable_tutorial.states import DataTableState, DataTableState2

```

## Adding Interactivity to our DataTable

Now we will add interactivity to our datatable. We do this using event handlers and event triggers.

> For our first interactivity example we will use the `on_cell_clicked` event trigger, this is called when the user clicks on a cell of the data editor. The event trigger receives the coordinates of the cell.

Reviewer: ```suggestion
The first example implements a handler for the `on_cell_clicked` event trigger, which is called when the user clicks on a cell of the data editor. The event trigger receives the coordinates of the cell.
```

```python
class DataTableState(rx.State):
    clicked_cell: str = "Cell clicked: "

    ...

>     def get_clicked_data(self, pos) -> str:

Reviewer: ```suggestion
    def get_clicked_data(self, pos: tuple[int, int]) -> str:
```

Best to add an annotation here for readability, otherwise i have no idea what this `pos` object looks like

Author: added

        self.clicked_cell = f"Cell clicked: \{pos}"
        
```

> Here we define a state var called `clicked_cell` that we will use to keep track of the clicked cell. We define an event handler `get_clicked_data` that updates the value of the `clicked_cell` var when it is called. In essence, we have clicked on a cell, called the `on_cell_clicked` event trigger which calls the `get_clicked_data` event handler, which updates the `clicked_cell` var.

Reviewer: ```suggestion
The state has a var called `clicked_cell` that will store a message about which cell was clicked. We define an event handler `get_clicked_data` that updates the value of the `clicked_cell` var when it is called. In essence, we have clicked on a cell, called the `on_cell_clicked` event trigger which calls the `get_clicked_data` event handler, which updates the `clicked_cell` var.
```

```python demo
rx.heading(DataTableState.clicked_cell)
```

```python demo
rx.data_editor(
    columns=DataTableState.cols,
    data=DataTableState.data,
    on_cell_clicked=DataTableState.get_clicked_data,
)
```

> The event handler `on_cell_context_menu` can be used in the same way as `on_cell_clicked`, except here the event trigger is called when the user right clicks, i.e. when the cell should show a context menu.

Reviewer: Do we have any example anywhere of displaying a custom context menu? I feel like this is a bit of a tease if we're not going to show the code. And it's a bit more complicated than it sounds here.

At the end of the day, we might leave this line out, or save it for later in the tutorial/API reference.

Author: We give an example of on_group_header_context_menu later on. These are not really custom context menus, it is just an event handler that picks up a right mouse click

### Editing cells

> Another important type of interactivity we will showcase is how to edit cells. Here we use the `on_cell_edited` event trigger.

Reviewer: ```suggestion
Another important type of interactivity we will showcase is how to edit cells. Here we use the `on_cell_edited` event trigger to update the data based on what the user entered.
```

```python
class DataTableState(rx.State):
    clicked_cell: str = "Cell clicked: "
    edited_cell: str = "Cell edited: "

    ...

    def get_clicked_data(self, pos) -> str:
        self.clicked_cell = f"Cell clicked: \{pos}"

>     def get_edited_data(self, pos, val) -> str:

Reviewer: Suggest more type annotations here.

What is the type of `val`? Is it based on the column type for the table? Is it always a `str` and I have to cast it myself?

Author: val is a dictionary like this: {'kind': 'text', 'data': '20', 'displayData': 'Hermione Jean Granger', 'readonly': False, 'allowOverlay': True} and we index in and just use the 'data' key for now

        col, row = pos
        self.data[row][col] = val["data"]
        self.edited_cell = f"Cell edited: \{pos}, Cell value: \{val['data']}"
        
```

The `on_cell_edited` event trigger is called when the user modifies the content of a cell. It receives the coordinates of the cell and the modified content. We pass these into the `get_edited_data` event handler and use them to update the `data` state var at the appropriate position. We then update the `edited_cell` var value.

```python demo

> rx.heading(DataTableState.edited_cell)

Reviewer: might not want to use `heading` here, as these are bigger than the page subheadings and make skimming more difficult

Author: changed

        },
        {
            "title": "Human",
            "type": "bool",
            "group": "Data",
            "width": 80,
        },
        {
            "title": "House",
            "type": "str",
            "group": "Data",
        },
        {
            "title": "Wand",
            "type": "str",
            "group": "Data",
            "width": 250,
        },
        {
            "title": "Patronus",
            "type": "str",
            "group": "Data",
        },
        {
            "title": "Blood status",
            "type": "str",
            "group": "Data",
            "width": 200,
        },
    ]

    ...
```

The table now has a header as below.

```python demo
rx.data_editor(
    columns=DataTableState2.cols,
    data=DataTableState2.data,
    on_cell_clicked=DataTableState2.get_clicked_data,
    on_cell_edited=DataTableState2.get_edited_data,
)
```

There are several event triggers we can apply to the group header.

```python
class DataTableState2(rx.State):
    """The app state."""

    right_clicked_group_header : str = "Group header right clicked: "

    ...

    def get_group_header_right_click(self, index, val):

>         self.right_clicked_group_header = f"Group header right clicked at index: \{index}, Group header value: \{val['group']}"

Reviewer: not sure i see the utility in this example. it works, but i don't know why someone would want this

Author: it is just an example that shows how to do a right click event handler. I assume people will figure out their own uses for it if they want to ever create events based on right clicks from the user

```

```python demo
rx.heading(DataTableState2.right_clicked_group_header)
```

```python demo
rx.data_editor(
    columns=DataTableState2.cols,
    data=DataTableState2.data,
    on_cell_clicked=DataTableState2.get_clicked_data,
    on_cell_edited=DataTableState2.get_edited_data,
    on_group_header_context_menu=DataTableState2.get_group_header_right_click,
)
```

In this example we use the `on_group_header_context_menu` event trigger which is called when the user right-clicks on a group header. It returns the `index` and the `data` of the group header. We can also use the `on_group_header_clicked` and `on_group_header_renamed` event triggers which are called when the user left-clicks on a group header and when a user renames a group header respectively.

### More Event Triggers

> There are several other event triggers that are worth exploring. The `on_item_hovered` event trigger is called whenever the user hovers over an item in the datatable. The `on_delete` event trigger is called when the user deletes a cell from the datatable. 

Reviewer: > The `on_delete` event trigger is called when the user deletes a cell from the datatable. 

TIL that deleting a cell was even possible 😁

> The final event trigger to check out is `on_column_resize`, this allows us to manually resize the width of our columns by hand. The event trigger returns the `col` we are adjusting and the new `width` we have defined. We then index into `self.cols` defined in our state and change the `width` of that column using this code: `self.cols[col['pos']]['width'] = width`.

Reviewer: > this allows us to manually resize the width of our columns by hand

This sounds sort of funny to me. 

Maybe something more like, "on_column_resize allows us to respond to the user dragging the handle between columns"

I think more description of the `col` arg is warranted, since it appears to be a dictionary, but not clear what is inside.

Author: added both of these
